### PR TITLE
Issue #581 - use a repository id to identify 'real' maven dependencies

### PR DESCRIPTION
**Pull request by laeubi**
Signed-off-by: Christoph Läubrich <laeubi@laeubi-soft.de>

### CODE REVIEW

File: tycho-packaging-plugin/src/main/java/org/eclipse/tycho/packaging/AbstractTychoPackagingMojo.java [hunk 1]
```diff
@@ -172,6 +173,10 @@ protected MavenProject updatePom(String finalName) throws IOException {
 		for (Dependency dep : list) {
 			Dependency copy = dep.clone();
 			copy.setSystemPath(null);
+			if (dep.getGroupId().startsWith("p2.eclipse.") && Artifact.SCOPE_SYSTEM.equals(dep.getScope())) {
+				copy.setOptional(true);
+				copy.setScope(Artifact.SCOPE_PROVIDED);
+			}
```
Comment (mickaelistria) [inline tycho-packaging-plugin/src/main/java/org/eclipse/tycho/packaging/AbstractTychoPackagingMojo.java:179]: Isn't it/shouldn't it be part of a separate commit?
Couldn't it be then that instead of a repository id, we could then assume that everything that's not "provided" comes either from reactor or repository?

**Comment by mickaelistria** [comment]
As mentioned on another bug, I'm not sure we want to make the repository-id leak everywhere as it's not reliable (it's mutable, moveable, duplicat-able...) content. However, I think the scope is something we can rely on more safely.

File: tycho-packaging-plugin/src/main/java/org/eclipse/tycho/packaging/AbstractTychoPackagingMojo.java (hunk 1 above)
Comment (laeubi) [reply tycho-packaging-plugin/src/main/java/org/eclipse/tycho/packaging/AbstractTychoPackagingMojo.java:179]: > Isn't it/shouldn't it be part of a separate commit?
Maybe yes, I just want to make the big picture work here :-)

> Couldn't it be then that instead of a repository id, we could then assume that everything that's not "provided" comes either from reactor or repository?

I hope we can over the time make more and more items with the real scope, and maybe provided would become a valid scope then also even though we currently only use compile+test+system, so my goal is to be as specific as possible here.

**Comment by laeubi** [comment]
> As mentioned on another bug, I'm not sure we want to make the repository-id leak everywhere as it's not reliable (it's mutable, moveable, duplicat-able...) content. However, I think the scope is something we can rely on more safely.

The scope is nothing we can carry over in the p2-data as an artififact might be referenced in different projects with different scopes.

I agree that the repository-id is somewhat weak, but it at least indicates that the artifact was consumed one from a repository with that id (most often from `central` what is a default repository in maven and thus well known).

As mentioned in the comments, the longer goal would be to make this more configurable, so for example one might only want things converted that are from `central` or `nexus` or ... and as these ids match to the repository configuration of maven build, we can even query that repository for the artifact and skip if its not there.

**Comment by mickaelistria** [comment]
> The scope is nothing we can carry over in the p2-data as an artififact might be referenced in different projects with different scopes.

That's right, I got confused.

> I agree that the repository-id is somewhat weak, but it at least indicates that the artifact was consumed one from a repository with that id (most often from central what is a default repository in maven and thus well known).

OK

> As mentioned in the comments, the longer goal would be to make this more configurable, so for example one might only want things converted that are from central or nexus or ... and as these ids match to the repository configuration of maven build, we can even query that repository for the artifact and skip if its not there.

I really hope that we can keep Tycho simple enough so that people wouldn't need such fine grain configuration!

Anyway, I'm fine with this patch if you want to merge it, and I now acknowledge that storing repository-id is not perfect but not too bad either.

**Comment by laeubi** [comment]
> I really hope that we can keep Tycho simple enough so that people wouldn't need such fine grain configuration!

I use tycho with nearly zero configuration and actually that's the goal, its just that there are some cases where people need to comply to some "rules" so I'm always try to guess what *might* be necessary here in the future if some one demands it.

For me the important think is at the moment that we can express "this artifact was consumed at once from any remote repository and if you would add this your build will probably succeed in resolve it from there" versus "these artifact was probably never deployed at any maven repository at all" (beside the local one...).

**Comment by laeubi** [comment]
Great I hope the last integreation test will now also pass, that reminds me of 
- #106 

we still have some issues about confusing packageing types with artifact types...
